interfaces/apparmor/template.go

```diff
@@ -508,3 +508,164 @@ var nfsSnippet = `
+  # actual layout at a given moment.
+  /run/snapd/ns/snap.###SNAP_NAME###.fstab rw,
+  /run/snapd/ns/*.fstab.* rw,
+
+  # NOTE: at this stage the /snap directory is stable as we have called
+  # pivot_root already.
+
+  # Needed to perform mount/unmounts.
+  capability sys_admin,
+  # Needed for mimic construction.
+  capability chown,
+
+  # Allow freezing and thawing the per-snap cgroup freezers
+  /sys/fs/cgroup/freezer/snap.###SNAP_NAME###/freezer.state rw,
+
+  # Support mount profiles via the content interface. This should correspond
+  # to permutations of $SNAP -> $SNAP for reading and $SNAP_{DATA,COMMON} ->
+  # $SNAP_{DATA,COMMON} for both reading and writing.
+  #
+  # Note that:
+  #   /snap/*/*/**
+  # is meant to mean:
+  #   /snap/$SNAP_NAME/$SNAP_REVISION/and-any-subdirectory
+  # but:
+  #   /var/snap/*/**
+  # is meant to mean:
+  #   /var/snap/$SNAP_NAME/$SNAP_REVISION/
+  # NOTE: the wildcards below are there because content interface can do content
+  # sharing from one snap to another and we just don't know yet. Those should go
+  # away when the content interface starts using AddUpdateNS() snippets.
+  mount options=(ro bind) /snap/*/** -> /snap/*/*/**,
+  mount options=(ro bind) /snap/*/** -> /var/snap/*/**,
+  mount options=(rw bind) /var/snap/*/** -> /var/snap/*/**,
+  mount options=(ro bind) /var/snap/*/** -> /var/snap/*/**,
+
+  # Allow creating missing mount directories under $SNAP_DATA.
+  #
+  # The "tree" of permissions is needed for SecureMkdirAll that uses
+  # open(..., O_NOFOLLOW) and mkdirat() using the resulting file
+  # descriptor.
+  / r,
+  /var/ r,
+  /var/snap/ r,
+  /var/snap/###SNAP_NAME###/ r,
+  /var/snap/###SNAP_NAME###/** rw,
+
+  # Allow creating placeholder directory in /tmp/.snap/ as support for
+  # the writable mimic code that can poke holes in arbitrary read-only
+  # places using tmpfs and bind mounts.
+  /tmp/ r,
+  /tmp/.snap/{,**} rw,
+  # Allow mounting/unmounting any part of $SNAP over to a temporary place
+  # in /tmp/.snap/ during the preparation of a writable mimic.
+  # FIXME: update this with per-snap snap-update-ns profiles
+  mount options=(bind, rw) /** -> /tmp/.snap/**,
+  mount options=(rbind, rw) /** -> /tmp/.snap/**,
+  # Allow mounting tmpfs over the original read-only directory.
+  # FIXME: update this with per-snap snap-update-ns profiles
+  mount fstype=tmpfs options=(rw) tmpfs -> /**,
+  # Allow bind mounting anything from the temporary place in /tmp/.snap/
+  # back to $SNAP/** (to re-construct the data that was there before).
+  # FIXME: update this with per-snap snap-update-ns profiles
+  mount options=(bind, rw) /tmp/.snap/** -> /**,
+  mount options=(rbind, rw) /tmp/.snap/** -> /**,
+  # Allow unmounting the temporary directory in /tmp once it is no longer
+  # necessary.
+  umount /tmp/.snap/**,
+  # Allow unmounting any of the above in case something fails and
+  # we start recovery.
+  umount /**,
```

Comment:
Like I mentioned in the previous PR, this umount rules is a superset of the preceding. Can this be split into snap-specific umounts and snippet umounts?

Reply:
Yes, those will be adjusted when we have content interface and layout do the extra rules (I may pull this into one PR since you already looked and it's the next day :-)